### PR TITLE
Defer parsing the filename when viewing mail

### PR DESCRIPTION
**Pull request by romanofski**
This gives us a slight speed increase when populating the index for
results with lots of mails. Fixes issue #60.

Issue: #51

### CODE REVIEW

**Comment by romanofski** [comment]
Had to chuck away one test for this. Which means this stuff is untested at this point :/

File: src/Storage/ParsedMail.hs [hunk 1]
```diff
@@ -7,18 +7,21 @@ import Codec.MIME.Parse (parseMIMEMessage)
...
-        Left e -> pure $ Left $ show e
-        Right contents -> pure $ Right $ MIMEMail (parseMIMEMessage contents)
+parseMail :: NotmuchMail
+          -> FilePath
+          -> IO (Either String ParsedMail)
+parseMail m dbpath =
+    mailFilepath m dbpath
+           >>= either fail (\x -> try (readFile x):: IO (Either IOError T.Text))
+           >>= either (\x -> pure $ Left (show x <> " Maybe run notmuch new again?")) (pure . Right)
+           >>= either (pure . Left . show) (pure . Right . MIMEMail <$> parseMIMEMessage)
```
Comment (frasertweedale) [inline src/Storage/ParsedMail.hs]: `fail` is bad bad bad.

Do this instead:
```haskell
parseMail m dbpath =                                                                                     
  mailFilepath m dbpath                                                                                  
  >>= fmap (bimap (\x -> show x <> " Maybe run notmuch new again?") (MIMEMail . parseMIMEMessage) . join)
      . traverse (fmap (first (show :: IOError -> String)) . try . T.readFile)                           
```

The whole thing will improve once I do the "abstract over error type" refactor, though.

File: src/Storage/Notmuch.hs [hunk 2]
```diff
@@ -34,15 +34,34 @@ getMessages :: T.Text
...
+mailFilepath m dbpath =
+    withRODatabase dbpath go >>=
+    either
+        (pure . Left . show)
+        (pure .
+         maybe
+             (Left "No message found. Is your notmuch correctly setup?")
+             Right)
+  where
+    go db = findMessage db (view mailId m) >>= traverse messageFilename
```
Comment (frasertweedale) [inline src/Storage/Notmuch.hs]: can tidy this one up a bit:
```haskell
mailFilepath :: NotmuchMail -> FilePath -> IO (Either String FilePath)              
mailFilepath m dbpath =                                                             
    join . bimap show (maybe (Left notFoundMsg) Right) <$> withRODatabase dbpath go 
  where                                                                             
    notFoundMsg = "No message found. Is your notmuch correctly setup?"              
    go db = findMessage db (view mailId m) >>= traverse messageFilename             
```

Again, the forthcoming refactoring should improve things a bit more.

**Comment by frasertweedale** [comment]
@romanofski can we get a proper rebase and conflict resolution for this PR?

**Comment by romanofski** [comment]
hm... what the hell...

**Comment by frasertweedale** [comment]
@romanofski I think this is good to do but can you please build and run it as a sanity check? :)

Seriously I never run the program, I just refactor stuff and send PRs :)

**Comment by romanofski** [comment]
Yep working fine. There is already a test which covers viewing mail, so in case something would have gone south you would have noticed it. With the test re-factor I'll have the possibility to add in a test which also checks the error handling. At least the case in which the mail file is gone, but it's still indexed.

**Pull request merged**